sql/core/src/main/scala/org/apache/spark/sql/UDFRegistration.scala

```diff
@@ -1121,3 +1123,17 @@ class UDFRegistration private[sql] (functionRegistry: FunctionRegistry) extends
   // scalastyle:on line.size.limit
 
 }
+
+private[sql] object UDFRegistration {
+  /**
+   * Obtaining the schema of output encoder for `ScalaUDF`.
+   *
+   * As the serialization in `ScalaUDF` is for individual column, not the whole row,
+   * we just take the data type of vanilla object serializer, not `serializer` which
+   * is transformed somehow for top-level row.
+   */
+  def outputSchema(outputEncoder: ExpressionEncoder[_]): ScalaReflection.Schema = {
```

Comment:
shall we define it in `ExpressionEncoder`?

Reply:
I moved it out per https://github.com/apache/spark/pull/31103#discussion_r554405600. Do you think it is better to put in `ExpressionEncoder`?

Reply:
ah makes sense, it's udf registration specific. Let's keep it here.